core/storage/couchStorage.go

```diff
@@ -0,0 +1,1801 @@
+		}
+		existingObject = nil
+	}
+
+	if existingObject != nil {
+		if (metaData.DestinationPolicy != nil && existingObject.MetaData.DestinationPolicy == nil) ||
+			(metaData.DestinationPolicy == nil && existingObject.MetaData.DestinationPolicy != nil) {
+			return nil, &common.InvalidRequest{Message: "Can't update the existence of Destination Policy"}
+		}
+
+		if metaData.MetaOnly {
+			metaData.DataID = existingObject.MetaData.DataID
+			metaData.ObjectSize = existingObject.MetaData.ObjectSize
+			metaData.ChunkSize = existingObject.MetaData.ChunkSize
+		}
+		if metaData.DestinationPolicy != nil {
+			dests = existingObject.Destinations
+		}
+	}
+
+	newObject := couchObject{ID: id, MetaData: metaData, Status: status, PolicyReceived: false,
+		RemainingConsumers: metaData.ExpectedConsumers,
+		RemainingReceivers: metaData.ExpectedConsumers, Destinations: dests}
+
+	// In case of existing object, check if it had attachment and add to newObject if present
+	// This is done only in case of metaOnly update
+	// otherwise updated attachment will be added in the next block
+	if existingObject != nil {
+		newObject.Rev = existingObject.Rev
+		if metaData.MetaOnly && data == nil {
+			attachment, err := store.getAttachment(id)
+			if err != nil {
+				if err != notFound {
+					return nil, err
+				}
+			} else {
+				attachments := &kivik.Attachments{id: attachment}
+				newObject.Attachments = attachments
+			}
+		}
+	}
+
+	// Add attachment to newObject for NoData=false
+	if !metaData.NoData && data != nil {
+		content := ioutil.NopCloser(bytes.NewReader(data))
+		defer content.Close()
+
+		attachment := &kivik.Attachment{ContentType: "application/octet-stream", Content: content}
+		attachments := &kivik.Attachments{id: attachment}
+		newObject.Attachments = attachments
+	}
+
+	// Cases where attachment needs to be removed are handled implicitly
+	// by not adding the existing attachment to newObject
+	if err := store.upsertObject(id, newObject); err != nil {
+		return nil, &Error{fmt.Sprintf("Failed to store an object. Error: %s.", err)}
+	}
+
+	return deletedDests, nil
+}
+
+// RetrieveObjectAndStatus returns the object meta data and status with the specified parameters
+func (store *CouchStorage) RetrieveObjectAndStatus(orgID string, objectType string, objectID string) (*common.MetaData, string, common.SyncServiceError) {
+	result := couchObject{}
+	id := createObjectCollectionID(orgID, objectType, objectID)
+
+	if err := store.getOne(id, &result); err != nil {
+		if err != notFound {
+			return nil, "", &Error{fmt.Sprintf("Failed to retrieve object's status. Error: %s.", err)}
+		}
+		return nil, "", nil
+	}
+	return &result.MetaData, result.Status, nil
+}
+
+// DeleteStoredObject deletes the object
+func (store *CouchStorage) DeleteStoredObject(orgID string, objectType string, objectID string) common.SyncServiceError {
+	id := createObjectCollectionID(orgID, objectType, objectID)
+	if err := store.deleteObject(id); err != nil {
```

Comment:
DeleteStoredObject function needs to delete **BOTH** object and object data. I think you also need to either call deleteAttachment func in deleteObject function (couchDBHelper) or here inside DeleteStoredObject function (couchStorage)
See [deleteObject func](https://github.com/open-horizon/edge-sync-service/blob/dde1bef314097c720c0026c127f4854b1cc82ee7/core/storage/mongoStorageHelpers.go#L65) in mongoStorageHelpers

Reply:
In both MongoDB and Bolt, the object data associated with an object is stored as a separate file, so it needs to be deleted separately while deleting the object. But in CouchDB, object data is stored as an attachment to the object in the same CouchDB document. So when an object is deleted, its associated attachment/data is deleted automatically. Should I add a comment here to avoid confusion in the future?

Reply:
Ohk I see. Yes please add a comment